crates/propolis-api-types/src/instance_spec/mod.rs

```diff
@@ -155,18 +155,152 @@
 
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
+use serde_with::{DeserializeFromStr, SerializeDisplay};
 
 pub use propolis_types::{CpuidIdent, CpuidValues, CpuidVendor, PciPath};
+use uuid::Uuid;
 
 pub mod components;
 pub mod v0;
 
-/// Type alias for keys in the instance spec's maps.
-type SpecKey = String;
+/// A key identifying a component in an instance spec.
+//
+// Some of the components Omicron attaches to Propolis VMs, like network
+// interfaces and Crucible disks, are described by database records with UUID
+// primary keys. It's natural to reuse these UUIDs as component identifiers in
+// Propolis, especially because it lets Omicron functions that need to identify
+// a specific component (e.g. a specific Crucible backend that should handle a
+// disk snapshot request) pass that component's ID directly to Propolis.
+//
+// In some cases it's not desirable or possible to use UUIDs this way:
+//
+// - Some components (like the cloud-init disk) don't have their own rows in the
+//   database and so don't have obvious UUIDs to use.
+// - Some objects (like Crucible disks) require both a device and a backend
+//   component in the spec, and these can't share the same key.
+// - Propolis users outside the control plane may not have any component UUIDs
+//   at all and may just want to use strings to identify all their components.
+//
+// For these reasons, the key type is a union of a UUID and a String. This
```

Comment:
suuuuuuper pedantic, but i wouldn't call this a "union", since Rust [also has `union` types](https://doc.rust-lang.org/reference/items/unions.html) which are untagged unions.

Reply:
Fixed in f8ecc34.